src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1260 @@
+    if (reg < APLIC_NUM_SETIx_REGS){
+        new_val &= vaplic->active[reg];
+        vaplic->ie[reg] &= ~(new_val);
+        new_val &= vaplic->hw[reg];
+        aplic_clr32_enbl(reg, new_val); 
+        // Alternative code. Waiting review.
+        // for(size_t i = (reg*APLIC_NUM_INTP_PER_REG); 
+        //     i < (reg*APLIC_NUM_INTP_PER_REG) + APLIC_NUM_INTP_PER_REG; i++){
+        //     if (vaplic_get_active(vcpu, i)){
+        //         if(vaplic_get_hw(vcpu,i)){
+        //             aplic_clr_enbl(i);
+        //         }
+        //         BIT32_CLR_INTP(vaplic->ie, i);
+        //         vaplic_update_hart_line(vcpu, GET_HART_INDEX(vcpu, i));
+        //     }
+        // }
+    }
+    spin_unlock(&vaplic->lock);
+}
+
+/**
+ * @brief Clear the enabled bit for a given interrupt
+ * 
+ * @param vcpu virtual cpu
+ * @param new_val interrupt to clear the enable bit
+ */
+static void vaplic_set_clrienum(struct vcpu *vcpu, uint32_t new_val){
+    struct vaplic *vaplic = &vcpu->vm->arch.vaplic;
+
+    spin_lock(&vaplic->lock);
+    if (vaplic_get_active(vcpu, new_val) &&
+        vaplic_get_enbl(vcpu, new_val)) {
+        if(vaplic_get_hw(vcpu, new_val)){
+            aplic_clr_enbl(new_val);
+        }
+        BIT32_CLR_INTP(vaplic->ie, new_val);
+        vaplic_update_hart_line(vcpu, GET_HART_INDEX(vcpu, new_val));
+    }
+    spin_unlock(&vaplic->lock);
+}
+
+/**
+ * @brief Write to target register of a given interrupt
+ * 
+ * @param vcpu virtual cpu
+ * @param intp_id interrupt ID
+ * @param new_val value to write to target
+ */
+static void vaplic_set_target(struct vcpu *vcpu, irqid_t intp_id, uint32_t new_val){
+    struct vaplic *vaplic = &vcpu->vm->arch.vaplic;
+    uint16_t hart_index = (new_val >> APLIC_TARGET_HART_IDX_SHIFT) & APLIC_TARGET_HART_IDX_MASK;
+    cpuid_t pcpu_id = vm_translate_to_pcpuid(vcpu->vm, hart_index);
+    uint32_t new_aplic_target = 0;
+    uint32_t new_vaplic_target = 0;
+
+    spin_lock(&vaplic->lock);
+    if(pcpu_id == INVALID_CPUID){
+        /** If the hart index is invalid, make it vcpu = 0 
+         *  and read the new pcpu.
+         *  Software should not write anything other than legal 
+         *  values to such a field */
+        hart_index = 0;
+        pcpu_id = vm_translate_to_pcpuid(vcpu->vm, hart_index);
+    }
+    
+    /** If prio is 0, set to 1 (max) according to the spec*/
+    new_val &= APLIC_TARGET_IPRIO_MASK;
+    if (new_val == 0) {
+        new_val = APLIC_TARGET_PRIO_DEFAULT;
+    }
+    /** Write the target CPU in hart index */
+    new_aplic_target = new_val|(pcpu_id << APLIC_TARGET_HART_IDX_SHIFT);
+    new_vaplic_target = new_val|(hart_index << APLIC_TARGET_HART_IDX_SHIFT);
+
+    if (vaplic_get_active(vcpu, intp_id) && 
+        vaplic_get_target(vcpu, intp_id) != new_val) {
+        if(vaplic_get_hw(vcpu, intp_id)){
+            aplic_set_target(intp_id, new_aplic_target);
+            if(aplic_get_target(intp_id) == new_aplic_target){
```

Comment:
Initially, my idea was to check it because of the priority field. But now that I look at it again, I realize that this did not fulfil the purpose anyway. Because if the priority was different, I wouldn't write in the virtual and then there would be inconsistency in the hart index field.

Reply:
I see that your comment below already touches on this problem, so I will fix it in the same commit.